Disk grains should not show virtual devices fix issue 57612

piterpunk opened the pull request:
### What does this PR do?
Changes the disks.py to populate "disks" and "SSDs" only with real devices, skipping virtual devices like devmaps, loopback and RAM disks.

### What issues does this PR fix or reference?
Fixes: #57612 

### Previous Behavior
In a machine with only one disk the grains shows:

```
# salt-call grains.get disks
local:
    - loop1
    - dm-1
    - loop6
    - dm-6
    - loop4
    - dm-4
    - loop2
    - dm-2
    - loop0
    - dm-0
    - loop7
    - dm-7
    - sda
    - loop5
    - dm-5
    - loop3
    - dm-3
# salt-call grains.get SSDs
local:
    - ram2
    - ram0
    - ram9
    - ram14
    - ram7
    - ram12
    - ram5
    - ram10
    - ram3
    - ram1
    - ram15
    - ram8
    - ram13
    - ram6
    - ram11
    - ram4
```

### New Behavior
Now it shows only the real device:

```
# salt-call grains.get SSDs
local:
# salt-call grains.get disks
local:
    - sda
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
[salt-runtests-20200609201317.log](https://github.com/saltstack/salt/files/4755812/salt-runtests-20200609201317.log)

### Commits signed with GPG?
No

Commit:
Fix grains/disks.py to show only the real disks

- The _linux_disks includes virtual block devices like RAM disks,
  loopback and devmaps in "disks" and "SSDs" grains.
- Changed to ignore any disk in the /sys/devices/virtual subsystem.

piterpunk commented:
Also fixes #51898

waynew reviewed (changes requested):
Hey @piterpunk thanks for the PR!

This looks great - just needs a couple of tests here to prevent regressions. If you need some help writing tests, I'd be happy to help out - I run test clinics [on Twitch](https://www.twitch.tv/saltstackinc/schedule) on Tuesday and Thursday. You can also ping me in #salt on freenode or on the community Slack.

Let me know if you have any questions!

Commit:
Added unit tests for disks._linux_disks

- Changed the classname to DisksGrainsTestCase. Previously it was
  IscsiGrainsTestCase, but no code related to iscsi is here.
- Added the test__linux_disks to test disks._linux_disks. To do a
  faithful test, it's included an array with data from many real block
  devices: SATA, Fiber Channel, VirtIO, ASM, Loopback, LVM and Ramdisk

waynew commented on tests/unit/grains/test_disks.py:
```diff
@@ -83,3 +83,43 @@ def test__windows_disks_retcode(self):
+            "1",
+            "1",
+        ]
+        with patch("glob.glob", MagicMock(return_value=files)), patch(
+            "salt.utils.path.readlink", MagicMock(side_effect=links)
+        ), patch("salt.utils.files.fopen", mock_open(read_data=contents)):
```
```suggestion
       patch_glob = patch("glob.glob", autospec=True, return_value=files)
       patch_readlink = patch("salt.utils.path.readlink", autospec=True, side_effect=links)
       patch_fopen = patch("salt.utils.files.fopen", mock_open(read_data=contents), autospec=True)
       with patch_glob, patch_readlink, patch_fopen:
```

I'm pretty sure that's the right syntax. Anyway, the most important part of the change is `autospec=True` - basically what that does is ensure that the patched thing looks/behaves like the thing it's patching - so if the underlying API changes then this test will also (correctly) break. IIRC, patch will produce a MagicMock by default - you can just give it a return_value or side_effect.

waynew reviewed (changes requested):
Hey thanks for the update! I mentioned in IRC, but here's the changes.

You can actually pre-define your patches and then use them in the `with` block. If you don't like doing that, that's OK. But definitely add `autospec=True`.

Otherwise, this looks great, thanks!

Commit:
Added autospec=True to test_disks.py unit test

- Added autospec=True to "patch" instructions
- Extracted patches from "with" block and pre-defined them to improve
  the code readability

piterpunk commented:
Done. Only mock_open doesn't accept autospec=True, which makes sense, since mock_open() already explicit mocks the open() call.

Thanks with the help! I'll use the lessons learned in next tests.

waynew reviewed (approved):
🚀 Looks great! Thanks for the changes 👍 

We're currently in code freeze, so expect a couple of weeks before this gets merged (post 3001.1 release)

waynew reviewed (changes requested):
Ah whoops - spoke a moment too soon! Can you also add a [changelog entry?](https://docs.saltstack.com/en/latest/topics/development/changelog.html)

*then* we'll be good 👍

Commit:
Added the Changelog entry for fix 57612

  - Fixes issue with virtual block devices, like loopbacks and LVMs,
    wrongly populating the "disks" or "SSDs" grains.

waynew reviewed (dismissed):
:rocket:

remichristiaan commented:
A question not related to this fix in particular but to the naming of the grain.
  
[Salt doc states](https://docs.saltstack.com/en/latest/topics/grains/): _Grains resolve to lowercase letters. For example, FOO, and foo target the same grain._

The grain SSDs is not lowercase (all the other standard grains are) and is not usable as lowercase .. e.g. 
`salt-call grains.item SSDs`  works but `salt-call grains.item ssds` not. 

Can this be changed so that it complies with what the docs state?

piterpunk commented:
> A question not related to this fix in particular but to the naming of the grain.
>
> The grain SSDs is not lowercase (all the other standard grains are) and is not usable as lowercase .. e.g. `salt-call grains.item SSDs` works but `salt-call grains.item ssds` not.
> 
> Can this be changed so that it complies with what the docs state?

I can change the "SSDs" grain name to "ssds", the grain is not used by any Salt module or state, I did a quick check on saltstack-formulas and looks it's unused there too.

Should I do the change?

DmitryKuzmenko commented:
@remichristiaan good catch. Thank you!
@piterpunk yes please. This sounds correct to have it lowercase accordingly to the docs.

Commit:
Changes the 'SSDs' grain name to 'ssds'

- By the documentation (https://docs.saltstack.com/en/latest/topics/grains),
  all grains needs to be resolved in lowered case, so the 'SSDs' grain was
  changed to 'ssds'.
